src/operator/setup.rs

```diff
@@ -0,0 +1,185 @@
+use crate::test::prelude::{Node, Pod, TestKubeClient};
+
+use anyhow::{anyhow, Result};
+use kube::api::ObjectList;
+use kube::{Resource, ResourceExt};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use stackable_operator::status::Conditions;
+use std::fmt::Debug;
+use std::thread;
+use std::time::{Duration, Instant};
+
+/// A Wrapper to avoid passing in client or cluster everywhere.
+pub struct TestCluster<T> {
+    client: TestKubeClient,
+    cluster: Option<T>,
+    options: TestClusterOptions,
+    timeouts: TestClusterTimeouts,
+}
+
+/// Some reoccurring common test cluster options.
+pub struct TestClusterOptions {
+    pub cluster_ready_condition_type: String,
+    pub pod_name_label: String,
+}
+
+/// Some reoccurring common test cluster timeouts.
+pub struct TestClusterTimeouts {
+    pub cluster_ready: Duration,
+    pub pods_terminated: Duration,
+}
+
+impl<T> TestCluster<T>
+where
+    T: Clone + Debug + DeserializeOwned + Resource<DynamicType = ()> + Serialize,
+{
+    /// This creates a kube client and should be executed at the start of every test.
+    pub fn new(options: TestClusterOptions, timeouts: TestClusterTimeouts) -> Self {
+        TestCluster {
+            client: TestKubeClient::new(),
+            cluster: None,
+            options,
+            timeouts,
+        }
+    }
+
+    /// Applies a custom resource, stores the returned cluster object and sleeps for
+    /// to seconds to give the operator time to react on the custom resource.
+    /// Without the sleep it can happen that tests run without any pods being created.
+    fn apply(&mut self, cluster: &T) -> Result<()> {
+        self.cluster = Some(self.client.apply(&serde_yaml::to_string(cluster)?));
+
+        // we wait here to give the operator time to react to the custom resource
```

Comment:
Yeah i think the sleep is an artifact here. Check wait_ready, there i check for status conditions in the custom resource if the installation is complete. At some point i added the check for the number of pods. Before that check and without the sleep the test just ran through immediately (because there was no status set yet). I try to remove it.